Mage.Sets/src/mage/sets/MarchOfTheMachine.java

```diff
@@ -513,8 +520,143 @@ protected void addSpecialCards(List<Card> booster, int number) {
+
+    private final BoosterStructure AABBBCCC = new BoosterStructure(
+            commonA, commonA,
+            commonB, commonB, commonB,
+            commonC, commonC, commonC
+    );
+    private final BoosterStructure AAABBCCC = new BoosterStructure(
+            commonA, commonA, commonA,
+            commonB, commonB,
+            commonC, commonC, commonC
+    );
+    private final BoosterStructure AAABBBCC = new BoosterStructure(
+            commonA, commonA, commonA,
+            commonB, commonB, commonB,
+            commonC, commonC
+    );
+
+    private final BoosterStructure UU_RUU = new BoosterStructure(
+            uncommonT, uncommonD, rare, uncommon, uncommon
+    );
+    private final BoosterStructure UR_UUU = new BoosterStructure(
+            uncommonT, rareD, uncommon, uncommon, uncommon
+    );
+    private final BoosterStructure RU_UUU = new BoosterStructure(
+            rareT, uncommonD, uncommon, uncommon, uncommon
+    );
+
+    private final BoosterStructure A1 = new BoosterStructure(archive);
+
+    // In order for equal numbers of each common to exist, the average booster must contain:
+    // 2.12 A commons (224 / 101)
+    // 2.93 B commons (296 / 101)
+    // 2.85 C commons (288 / 101)
+    private final RarityConfiguration commonRuns = new RarityConfiguration(
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC,
+            AABBBCCC, AABBBCCC, AABBBCCC, AABBBCCC, 
+            AAABBCCC, AAABBCCC, AAABBCCC, AAABBCCC, AAABBCCC,
+            AAABBCCC, AAABBCCC,
+            AAABBBCC, AAABBBCC, AAABBBCC, AAABBBCC, AAABBBCC,
+            AAABBBCC, AAABBBCC, AAABBBCC, AAABBBCC, AAABBBCC,
+            AAABBBCC, AAABBBCC, AAABBBCC, AAABBBCC, AAABBBCC
+    );
+
+    private final RarityConfiguration uncommonRuns = new RarityConfiguration(
```

Comment:
How did you get these ratios?

Looks like rare/mythic battles are not opened enough and rare/mythic DFCs are opened too much.

<details>

March of the Machine - boosters opened: 100000. Found cards: 1500000

C: 116 unique, min 4860, max 8188, total 890585
U: 100 unique, min 3255, max 4732, total 426934
R: 89 unique, min 744, max 2645, total 110557
M: 34 unique, min 371, max 1299, total 21812
L: 5 unique, min 9955, max 10120, total 50112

MOM L Swamp: 10120
MOM L Plains: 10101
MOM L Island: 9975
MOM L Mountain: 9961
MOM L Forest: 9955
MOM C Aetherblade Agent: 8188
MOM C Tarkir Duneshaper: 8182
MOM C Assimilate Essence: 8167
MOM C Pyretic Prankster: 8138
MOM C Order of the Mirror: 8104
MOM C War-Trained Slasher: 8099
MOM C Bonded Herdbeast: 8085
MOM C Deadly Derision: 8082
MOM C Chomping Kavu: 8064
MOM C Bola Slinger: 8063
MOM C Thunderhead Squadron: 8062
MOM C Hangar Scrounger: 8058
MOM C Coming In Hot: 8057
MOM C Ichor Shade: 8047
MOM C Swordsworn Cavalier: 8045
MOM C Arachnoid Adaptation: 8045
MOM C Golden-Scale Aeronaut: 8043
MOM C Moment of Truth: 8029
MOM C Tenured Oilcaster: 8025
MOM C Etched Host Doombringer: 8025
MOM C Zhalfirin Shapecraft: 8016
MOM C Trailblazing Historian: 8010
MOM C Iridescent Blademaster: 8007
MOM C Timberland Ancient: 7997
MOM C Final Flourish: 7995
MOM C Wary Thespian: 7993
MOM C Skittering Surveyor: 7991
MOM C Dreg Recycler: 7989
MOM C Sigiled Sentinel: 7986
MOM C Enduring Bondwarden: 7986
MOM C Burning Sun's Fury: 7985
MOM C Ichor Drinker: 7983
MOM C Aerial Boost: 7983
MOM C Kor Halberd: 7981
MOM C Stasis Field: 7975
MOM C Atraxa's Fall: 7975
MOM C Knight of the New Coalition: 7968
MOM C Fertilid's Favor: 7968
MOM C Oculus Whelp: 7966
MOM C Halo-Charged Skaab: 7963
MOM C Corrupted Conviction: 7951
MOM C Tidal Terror: 7951
MOM C Sunder the Gateway: 7951
MOM C Eyes of Gitaxias: 7950
MOM C Flitting Guerrilla: 7949
MOM C Seed of Hope: 7942
MOM C War Historian: 7940
MOM C Consuming Aetherborn: 7938
MOM C Unseal the Necropolis: 7938
MOM C Onakke Javelineer: 7937
MOM C Preening Champion: 7936
MOM C Searing Barb: 7936
MOM C Placid Rottentail: 7934
MOM C Volcanic Spite: 7934
MOM C Protocol Knight: 7933
MOM C Failed Conversion: 7932
MOM C Serpent-Blade Assailant: 7932
MOM C Vengeant Earth: 7932
MOM C Expedition Lookout: 7931
MOM C Phyrexian Archivist: 7924
MOM C Realmbreaker's Grasp: 7916
MOM C Cut Short: 7914
MOM C Mirran Banesplitter: 7914
MOM C Flywheel Racer: 7910
MOM C Thrashing Frontliner: 7903
MOM C Attentive Skywarden: 7902
MOM C Infected Defector: 7902
MOM C Wrenn's Resolve: 7900
MOM C Gloomfang Mauler: 7894
MOM C Nezumi Informant: 7894
MOM C Beamtown Beatstick: 7891
MOM C Negate: 7890
MOM C Urn of Godfire: 7887
MOM C Bladed Battle-Fan: 7880
MOM C Kithkin Billyrider: 7876
MOM C Kitesail: 7872
MOM C Ral's Reinforcements: 7871
MOM C Overgrown Pest: 7870
MOM C Cosmic Hunger: 7869
MOM C Portent Tracker: 7865
MOM C Vanquish the Weak: 7863
MOM C Redcap Heelslasher: 7859
MOM C Blighted Burgeoning: 7857
MOM C Saiba Cryptomancer: 7855
MOM C Inspired Charge: 7842
MOM C Meeting of Minds: 7836
MOM C Mirrodin Avenged: 7836
MOM C Karsus Depthguard: 7835
MOM C Ephara's Dispersal: 7834
MOM C Etched Familiar: 7816
MOM C Wildwood Escort: 7813
MOM C Halo Hopper: 7810
MOM C Disturbing Conversion: 7808
MOM C Alabaster Host Sanctifier: 7804
MOM C Converter Beast: 7802
MOM C Akki Scrapchomper: 7790
MOM C Marauding Dreadship: 7781
MOM C Furtive Analyst: 7780
MOM C Scrollshift: 7779
MOM C Crystal Carapace: 7775
MOM C Angelic Intervention: 7757
MOM C Shatter the Source: 7757
MOM C Alabaster Host Intercessor: 7749
MOM C Temporal Cleansing: 7749
MOM C Furnace Host Charger: 7719
MOM C Traumatic Revelation: 7675
MOM C Tranquil Cove: 5138
MOM C Scoured Barrens: 5039
MOM C Rugged Highlands: 5034
MOM C Wind-Scarred Crag: 5029
MOM C Thornwood Falls: 4998
MOM C Jungle Hollow: 4980
MOM C Bloodfell Caves: 4977
MOM C Dismal Backwater: 4941
MOM C Blossoming Sands: 4892
MOM C Swiftwater Cliffs: 4860
MOM U Rampaging Geoderm: 4732
MOM U Elspeth's Smite: 4703
MOM U Botanical Brawler: 4685
MOM U Gift of Compleation: 4685
MOM U Artistic Refusal: 4678
MOM U Mirror-Shield Hoplite: 4672
MOM U Kami of Whispered Hopes: 4665
MOM U Mutagen Connoisseur: 4654
MOM U Sculpted Perfection: 4649
MOM U Merciless Repurposing: 4648
MOM U Compleated Huntmaster: 4648
MOM U Norn's Inquisitor: 4645
MOM U Phyrexian Awakening: 4643
MOM U Scrappy Bruiser: 4641
MOM U Marshal of Zhalfir: 4639
MOM U Copper Host Crusher: 4638
MOM U Stormclaw Rager: 4633
MOM U Wicked Slumber: 4631
MOM U Omen Hawker: 4630
MOM U Shivan Branch-Burner: 4629
MOM U Xerex Strobe-Knight: 4616
MOM U Sandstalker Moloch: 4616
MOM U Lithomantic Barrage: 4606
MOM U Change the Equation: 4603
MOM U Tangled Skyline: 4601
MOM U Fearless Skald: 4597
MOM U Storm the Seedcore: 4588
MOM U Stoke the Flames: 4587
MOM U Oracle of Tragedy: 4580
MOM U Surge of Salvation: 4577
MOM U Scorn-Blade Berserker: 4576
MOM U Elvish Vatkeeper: 4575
MOM U Streetwise Negotiator: 4568
MOM U Tandem Takedown: 4561
MOM U Glistening Deluge: 4559
MOM U Invasion of Vryn: 4555
MOM U Invasion of New Capenna: 4553
MOM U Invasion of Kamigawa: 4551
MOM U Astral Wingspan: 4551
MOM U Phyrexian Gargantua: 4547
MOM U Furnace Reins: 4547
MOM U Corruption of Towashi: 4546
MOM U Invasion of Azgol: 4544
MOM U Joyful Stormsculptor: 4542
MOM U Seal from Existence: 4532
MOM U Invasion of Moag: 4528
MOM U Ramosian Greatsword: 4526
MOM U Invasion of Mercadia: 4525
MOM U Collective Nightmare: 4524
MOM U Render Inert: 4523
MOM U Invasion of Xerex: 4522
MOM U Ravenous Sailback: 4516
MOM U Phyrexian Censor: 4513
MOM U Invasion of Eldraine: 4512
MOM U Invasion of Amonkhet: 4512
MOM U Zhalfirin Lancer: 4511
MOM U Furnace Gremlin: 4502
MOM U Tiller of Flesh: 4500
MOM U Invasion of Zendikar: 4485
MOM U Halo Forager: 4469
MOM U Invasion of Ergamon: 4447
MOM U Invasion of Dominaria: 4443
MOM U Invasion of Kylem: 4439
MOM U Invasion of Muraganda: 4436
MOM U Invasion of Pyrulea: 4419
MOM U Invasion of Ulgrotha: 4408
MOM U Invasion of Lorwyn: 4395
MOM U Invasion of Kaladesh: 4388
MOM U Invasion of Belenon: 4379
MOM U Invasion of Regatha: 4296
MOM U Nezumi Freewheeler: 4139
MOM U Gnottvold Hermit: 4086
MOM U Skyclave Aerialist: 4078
MOM U Khenra Spellspear: 4051
MOM U Sun-Blessed Guardian: 4015
MOM U Blightreaper Thallid: 3995
MOM U Seraph of New Capenna: 3957
MOM U Herbology Instructor: 3953
MOM U Harried Artisan: 3948
MOM U Captive Weird: 3937
MUL U Inga Rune-Eyes: 3524
MUL U Shanna, Sisay's Legacy: 3457
MUL U Juri, Master of the Revue: 3447
MUL U Daxos, Blessed by the Sun: 3444
MUL U Aegar, the Freezing Flame: 3439
MUL U Dina, Soul Steeper: 3426
MUL U Rona, Sheoldred's Faithful: 3415
MUL U Yargle, Glutton of Urborg: 3411
MUL U Renata, Called to the Hunt: 3389
MUL U Reyav, Master Smith: 3389
MUL U Tymaret, Chosen from Death: 3385
MUL U Raff, Weatherlight Stalwart: 3380
MUL U Radha, Coalition Warlord: 3351
MUL U Zada, Hedron Grinder: 3342
MUL U Fynn, the Fangbearer: 3341
MUL U Kwende, Pride of Femeref: 3339
MUL U Imoti, Celebrant of Bounty: 3332
MUL U Firja, Judge of Valor: 3287
MUL U Valduk, Keeper of the Flame: 3278
MUL U Tetsuko Umezawa, Fugitive: 3255
MOM R Polukranos Reborn: 2645
MOM R Heliod, the Radiant Dawn: 2618
MOM R Rona, Herald of Invasion: 2608
MOM R Etali, Primal Conqueror: 2502
MOM R Ayara, Widow of the Realm: 2469
MOM R Boon-Bringer Valkyrie: 1526
MOM R Chrome Host Seedshark: 1496
MOM R Glistening Dawn: 1489
MOM R Deeproot Wayfinder: 1482
MOM R Archpriest of Shadows: 1480
MOM R Nahiri's Warcrafting: 1480
MOM R Knight-Errant of Eos: 1477
MOM R Doomskar Warrior: 1476
MOM R Baral and Kari Zev: 1470
MOM R Breach the Multiverse: 1468
MOM R Pile On: 1467
MOM R Ancient Imperiosaur: 1461
MOM R Progenitor Exarch: 1455
MOM R See Double: 1454
MOM R Kogla and Yidaro: 1450
MOM R Errant and Giada: 1448
MOM R Glissa, Herald of Predation: 1446
MOM R Hidetsugu and Kairi: 1438
MOM R Djeru and Hazoret: 1437
MOM R Sunfall: 1437
MOM R Yargle and Multani: 1436
MOM R Hoarding Broodlord: 1435
MOM R Bloodfeather Phoenix: 1430
MOM R Inga and Esika: 1430
MOM R Ozolith, the Shattered Spire: 1428
MOM R Transcendent Message: 1426
MOM R Rampaging Raptor: 1424
MOM R Zephyr Singer: 1419
MOM R Tribute to the World Tree: 1416
MOM R Dusk Legion Duelist: 1413
MOM R Omnath, Locus of All: 1406
MOM R Guardian of Ghirapur: 1404
MOM R Voldaren Thrillseeker: 1404
MOM R Ghalta and Mavren: 1399
MOM R Into the Fire: 1397
MOM R Quintorius, Loremaster: 1391
MOM R Grafted Butcher: 1387
MOM R Bloated Processor: 1387
MOM R Realmbreaker, the Invasion Tree: 1386
MOM R Complete the Circuit: 1385
MOM R Faerie Mastermind: 1381
MOM R Drana and Linvala: 1364
MOM R Rankle and Torbran: 1356
MOM R City on Fire: 1352
MOM M Elesh Norn: 1299
MOM M Urabrask: 1284
MOM M Vorinclex: 1254
MOM M Sheoldred: 1250
MOM M Jin-Gitaxias: 1215
MUL R Lutri, the Spellchaser: 956
MUL R Judith, the Scourge Diva: 949
MUL R Seizan, Perverter of Truth: 939
MUL R Yedora, Grave Gardener: 936
MUL R Lathiel, the Bounteous Dawn: 934
MUL R Baral, Chief of Compliance: 919
MUL R Sram, Senior Edificer: 918
MUL R Kaheera, the Orphanguard: 916
MUL R Firesong and Sunspeaker: 913
MUL R Emry, Lurker of the Loch: 911
MUL R Arixmethes, Slumbering Isle: 910
MUL R Umori, the Collector: 907
MUL R Yorion, Sky Nomad: 907
MUL R Keruga, the Macrosage: 906
MUL R Captain Lannery Storm: 904
MUL R Ayara, First of Locthwain: 900
MUL R Obosh, the Preypiercer: 898
MUL R Lurrus of the Dream-Den: 896
MUL R Jegantha, the Wellspring: 877
MUL R Gyruda, Doom of Depths: 876
MUL R Goreclaw, Terror of Qal Sisma: 876
MUL R Zirda, the Dawnwaker: 875
MUL R Thalia, Guardian of Thraben: 874
MUL R Anafenza, Kin-Tree Spirit: 864
MUL R Atris, Oracle of Half-Truths: 861
MUL R Horobi, Death's Wail: 847
MUL R Squee, the Immortal: 846
MOM R Invasion of Ixalan: 846
MUL R Teysa Karlov: 844
MUL R Brudiclad, Telchor Engineer: 842
MOM R Invasion of Tolvada: 811
MOM R Invasion of Alara: 810
MOM R Invasion of Karsus: 806
MOM R Invasion of Ikoria: 803
MOM R Invasion of Segovia: 795
MOM R Invasion of Fiora: 794
MOM R Invasion of Gobakhan: 790
MOM M Zimone and Dina: 776
MOM R Invasion of Theros: 769
MOM R Invasion of Kaldheim: 753
MOM R Invasion of Arcavios: 744
MOM M Archangel Elspeth: 739
MOM M Kroxa and Kunoros: 733
MOM M Monastery Mentor: 723
MOM M Chandra, Hope's Beacon: 721
MOM M Borborygmos and Fblthp: 715
MOM M Thalia and The Gitrog Monster: 702
MOM M Zurgo and Ojutai: 701
MOM M Wrenn and Realmbreaker: 700
MOM M Sword of Once and Future: 690
MUL M Jin-Gitaxias, Core Augur: 501
MUL M Yarok, the Desecrated: 487
MUL M Urabrask the Hidden: 472
MUL M Elesh Norn, Grand Cenobite: 462
MUL M Skithiryx, the Blight Dragon: 459
MUL M Ezuri, Claw of Progress: 458
MUL M Aurelia, the Warleader: 457
MUL M Niv-Mizzet Reborn: 450
MUL M Vorinclex, Voice of Hunger: 448
MUL M Grimgrin, Corpse-Born: 440
MUL M Ragavan, Nimble Pilferer: 439
MUL M Atraxa, Praetors' Voice: 439
MUL M Sheoldred, Whispering One: 432
MOM M Invasion of Shandalar: 426
MUL M Kroxa, Titan of Death's Hunger: 424
MOM M Invasion of New Phyrexia: 399
MOM M Invasion of Ravnica: 373
MOM M Invasion of Innistrad: 373
MOM M Invasion of Tarkir: 371


</details>